lib/galaxy/tool_util/loader_directory.py

```diff
@@ -188,22 +188,41 @@ def looks_like_a_data_manager_xml(path):
     return looks_like_xml(path=path, regex=DATA_MANAGER_REGEX)
 
 
-def is_a_yaml_with_class(path, classes):
-    """Determine if a file is a valid YAML with a supplied ``class`` entry."""
-    if not _has_extension(path, YAML_EXTENSIONS):
-        return False
+def as_dict_if_looks_like_yaml_or_cwl_with_class(path, classes):
+    """
+    get a dict from yaml file if it contains `class: CLASS`, where CLASS is
+    any string given in CLASSES. must appear in the first 5k and also load
+    properly in total.
+    """
+    with open(path, encoding="utf-8") as f:
+        try:
+            start_contents = f.read(5 * 1024)
+        except UnicodeDecodeError:
+            return None
+        if re.search(rf"\nclass:\s+({'|'.join(classes)})\s*\n", start_contents) is None:
+            return None
 
     with open(path) as f:
         try:
             as_dict = yaml.safe_load(f)
         except Exception:
-            return False
+            return None
 
     if not isinstance(as_dict, dict):
-        return False
+        return None
 
     file_class = as_dict.get("class", None)
-    return file_class in classes
+    if file_class not in classes:
+        return None
+
+    return as_dict
```

Comment:
I don't think it's a good idea to overload the data with the result. Maybe return 2 values, one for the result looking like a document, the other with the dict or None.

Reply:
So your suggestion is to return 

- `False, None` or
- `True, as_dict`

?

Reply:
Yep, that'd work